aten/src/ATen/core/typeid.h

```diff
@@ -134,253 +237,123 @@ struct TypeNameRegisterer {
  */
 class AT_CORE_API TypeMeta {
  public:
-  using PlacementNew = void(void*, size_t);
-  using TypedCopy = void(const void*, void*, size_t);
-  using TypedDestructor = void(void*, size_t);
+  using PlacementNew = detail::TypeMetaData::PlacementNew;
+  using TypedCopy = detail::TypeMetaData::TypedCopy;
+  using TypedDestructor = detail::TypeMetaData::TypedDestructor;
+
   /** Create a dummy TypeMeta object. To create a TypeMeta object for a specific
    * type, use TypeMeta::Make<T>().
    */
-  TypeMeta() noexcept
-      : id_(TypeIdentifier::uninitialized()),
-        itemsize_(0),
-        ctor_(nullptr),
-        copy_(nullptr),
-        dtor_(nullptr) {}
+  constexpr TypeMeta() noexcept
+      : data_(&uninitialized_) {}
 
   /**
    * Copy constructor.
    */
-  TypeMeta(const TypeMeta& src) noexcept = default;
+  constexpr TypeMeta(const TypeMeta& src) noexcept = default;
 
   /**
    * Assignment operator.
    */
-  TypeMeta& operator=(const TypeMeta& src) noexcept = default;
+  AT_CPP14_CONSTEXPR TypeMeta& operator=(const TypeMeta& src) noexcept = default;
 
-  TypeMeta(TypeMeta&& rhs) noexcept = default;
+  constexpr TypeMeta(TypeMeta&& rhs) noexcept = default;
 
  private:
   // TypeMeta can only be created by Make, making sure that we do not
   // create incorrectly mixed up TypeMeta objects.
-  TypeMeta(
-      TypeIdentifier i,
-      size_t s,
-      PlacementNew* ctor,
-      TypedCopy* copy,
-      TypedDestructor* dtor) noexcept
-      : id_(i), itemsize_(s), ctor_(ctor), copy_(copy), dtor_(dtor) {}
-
-  // Mechanism for throwing errors which can't be prevented at compile time
-  // due to type erasure. E.g. somebody calling TypeMeta::copy() for
-  // non-copiable type. Right now just throws exception but is implemented
-  // in .cpp to manage dependencies
-  static void _ThrowRuntimeTypeLogicError(const std::string& msg);
+  constexpr TypeMeta(const detail::TypeMetaData* data) noexcept
+      : data_(data) {}
 
  public:
   /**
    * Returns the type id.
    */
-  const TypeIdentifier& id() const noexcept {
-    return id_;
+  constexpr TypeIdentifier id() const noexcept {
```

Comment:
if we ever decide to try microoptimizations - we can try to keep the id only in TypeMeta and have a global array to register stuff. But most likely it doesn't matter